### PR TITLE
Limit JAX GPU memory preallocation

### LINKED ISSUE #151: Ensure GPU Memory Allocation Request from JAX/Pytorch Are Reasonable
Tensorflow/Jax/Pytorch(?) can be aggressive with resource allocation. We should ensure that they don't ask for too much.

Relates to #149 

See https://docs.pytorch.org/docs/stable/cuda.html#torch.cuda.set_per_process_memory_fraction

Example: (test for cuda available needed)
```python
import torch
torch.cuda.set_per_process_memory_fraction(0.5)
```

``set_per_process_memory_fraction`` is a hard limit, so it doesn't seem like a great option since it will trigger an exception if memory usage goes above this fraction. More importantly, from what I've so far read, PyTorch's GPU memory allocation is not preallocated and is totally dynamic. In which case, there's not an issue there, though this should be empirically confirmed. It will, however, hold on memory for future use, which may need resolving using ``torch.cuda.empty_cache()``. For batch processing this may be a good thing and not needed.

See https://docs.jax.dev/en/latest/gpu_memory_allocation.html

### PR DESCRIPTION
**Pull request by jamienoss**
Resolves #151 
Relates to #169

Jax's default GPU memory preallocation consumption is 75%. This PR toggles this functionality using the env var ``XLA_PYTHON_CLIENT_PREALLOCATE``.

When testing this on our 80GB A100 with preallocation ON, the max memory footprint is ~60GB as expected. With it toggled OFF, the max memory footprint is ~8.6GB.

Jax docs:

- https://docs.jax.dev/en/latest/gpu_memory_allocation.html
- https://docs.jax.dev/en/latest/config_options.html

Extremely crude profiling confirmed that this works. Running and observing ``nvidia-smi -lms=10`` whilst also running ``python flfm/cli.py main /data/yale/light_field_image.tif /data/yale/measured_psf.tif reconstructed_image.tiff --normalize_psf=True --lens_radius=230 --lens_center="(1000,980)" --backend=jax``.

Max mem footprint observed with ``FLFM_GPU_MEMORY_PREALLOCATION=true``
Screenshot 2025-07-24 at 5 58 04 PM

> [!NOTE]  
>  ``81920 * 0.75 = 61440 ~= 61212`` as observered.

Max mem footprint observed with ``FLFM_GPU_MEMORY_PREALLOCATION=false``
Screenshot 2025-07-24 at 6 02 41 PM

### CODE REVIEW

File: flfm/backend/jax.py [hunk 1]
```diff
@@ -9,6 +10,10 @@
 from PIL import Image
 
 from flfm.backend.base import BaseIO, BaseRestoration
+from flfm.settings import settings
+
+# See https://docs.jax.dev/en/latest/gpu_memory_allocation.html
+os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = settings.GPU_MEMORY_PREALLOCATION
```
Comment (jamienoss) [inline flfm/backend/jax.py:16]: This may be happening too late to work.

**Comment by jamienoss** [comment]
@ryanhausen ``nsys`` wasn't installed on our boxes, so I at least ran the above sanity check, which seems good to me. I'd much prefer to have actual mem/time plots, though.

File: flfm/settings.py [hunk 2]
```diff
@@ -20,6 +20,7 @@ class Settings(BaseSettings):
     DEFAULT_CENTER_Y: int = 980
     DEFAULT_RADIUS: int = 230
     DEBUG_CLI: bool = False
+    GPU_MEMORY_PREALLOCATION: bool = False  # Only applicable for BACKEND = "jax".
```
[the remark this answers is not in the document]
Comment (jamienoss) [reply flfm/settings.py:23]: Yes, as noted both in the results of the description by toggling ``FLFM_ GPU_MEMORY_PREALLOCATION `` and as commented in https://github.com/ssec-jhu/flfm/pull/199#discussion_r2211639494.

**Review by ryanhausen — APPROVED** [review]
